rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,382 @@
+import collections
+import pkg_resources
+import socket
+import struct
+
+from .consts import SCPCommands, DataType
+from . import boot, consts
+from .scp_connection import SCPConnection
+from rig.utils.contexts import ContextMixin, Required
+
+
+class MachineController(ContextMixin):
+    """TODO
+
+    Attributes
+    ----------
+    """
+    def __init__(self, initial_host, n_tries=5, timeout=0.5, app_id=None,
+                 boot_data=None, struct_data=None):
+        """Create a new controller for a SpiNNaker machine.
+
+        Parameters
+        ----------
+        initial_host : string
+        n_tries : int
+        timeout : float
+        app_id : int or None
+        boot_data : bytes or None
+            Data to boot the machine with.  If None then the default boot file
+            provided with rig will be used.
+        struct_data : bytes or None
+            Data to interpret as a representation of the memory layout of data
+            within the memory of a running core.  If None then the default
+            struct file provided with rig will be used.
+        """
+        # Initialise the context stack
+        ContextMixin.__init__(self, {app_id: app_id})
+
+        # Store the initial parameters
+        self.initial_host = initial_host
+        self.n_tries = n_tries
+        self.timeout = timeout
+        self.app_id = app_id
+
+        # Store the struct and boot data
+        if struct_data is None:
+            struct_data = pkg_resources.resource_string("rig",
+                                                        "boot/sark.struct")
+        self.struct_data = struct_data
+
+        if boot_data is None:
+            boot_data = pkg_resources.resource_string("rig", "boot/scamp.boot")
+        self.boot_data = boot_data
+
+        # Create the initial connection
+        self.connections = [
+            SCPConnection(initial_host, n_tries, timeout)
+        ]
+
+    def __call__(self, **context_args):
+        """Create a new context for use with `with`.
+
+        E.g:
+            with controller(x=3, y=4):
+                # All commands will now communicate with chip (3, 4)
+        """
+        return self.get_new_context(**context_args)
+
+    def send_scp(self, x, y, *args, **kwargs):
```

Comment:
This is not trivial, because the context argument wrapping mucks around with `*args`.  I can continue faffing or I can cheat a little: do you have a preference?